feat: allow to opt-out of default babel plugins and presets

sastan opened the pull request:
The unique feature of generating typescript declaration files for all chunks is really great and we want (need) to use this plugin. 

But: we are using a own preset that internally configures `@babel/preset-env`. This can not be detected by `rollup-plugin-ts`. This PR introduces an option `noBabelConfigCustomization` to completly disable the default presets and plugins (which are a great feature!). This is opt-in or opt-out depending on the view ;-)

I hope this can be merged and happy to adjust this PR.

tunnckoCore commented:
Great! I'm for it too. This was causing problems in the past and in the TSDX try of integrating this plugin.

wessberg commented:
Hey there. Thanks for your contribution! Yes, I've been thinking along the same lines. So let's get the ball rolling on merging this PR 🙂.

One thought that I have is this: The plugin does some magic to separate minification-related plugins/presets (such as `babel-preset-minify` and related plugins) from the provided or discovered babel config(s) such that the majority of babel plugins run on a file-by-file basis except for minification-related plugins which then run on a chunk-by-chunk basis. The reason is that you get much better minification this way. And the reason for running everything else on a file-by-file basis is of course to make Rollup/Acorn understand whatever experimental syntax you may be transforming with babel.

It would be great that this continued to work, even if `rollup-plugin-ts` doesn't add any default presets/plugins. That would require some changes to your PR in `src/util/get-babel-config/get-babel-config.ts`.

I'm thinking along the lines of the following:
- Provide an option that can disable any default babel plugins/presets
- Move babel plugins and presets into optional dependencies (everything except `@babel/core` which is needed to resolve and manipulate config files).

How do you feel about that?

sastan commented:
### separate minification-related plugins/presets

*included in updated PR*

This is done using [`loadOptions`](https://babeljs.io/docs/en/babel-core#loadoptions) which returns a flat resolved list of plugins, that is then matched against the configured names.

As we do not know which plugins may get loaded for a filename `hasMinifyOptions` is always set to `true`. That currently results in some plugins applied twice (`renderChunk` and `transform`). Maybe we could change which plugins are selected during `renderChunk (`minifyConfig`). I would suggest to only include those that have been excluded during `transform` or are syntax plugins (`babel-plugin-syntax-*` or just `[-/]syntax-*`). What do you think?

### `modules` option of `preset-env` and the other yearly presets

*included in updated PR*

I included the `caller` api which is used by `preset-env` to detect if it can emit ES Modules. The user should use `modules: 'auto'` (the default value of `preset-env`). `modules: 'auto'` allows to use the same babel config for different transpile targets like rollup (esm) and jest (cjs) and it *just works*.

Alternatively `modules: false` would work aswell.

### Move babel plugins and presets into optional dependencies

*included in updated PR*

That would mean they get installed anyway. I think they should move into `peerDependencies` which would ensure only semver compatible versions are installed. Some package managers support [`peerDependenciesMeta`](https://github.com/yarnpkg/rfcs/blob/master/accepted/0000-optional-peer-dependencies.md) ([yarn](https://yarnpkg.com/en/docs/package-json#peerdependenciesmeta-), [pnpm](https://pnpm.js.org/en/package_json#peerdependenciesmeta-pkg-optional) and [npm@ ^6.11.0](https://github.com/npm/cli/pull/224) ) which allow users to **not** install those `peerDependencies`.

```json
{
  "peerDependencies": {
    "@babel/plugin-transform-runtime": "^7.6.2"
  },
  "peerDependenciesMeta": {
    "@babel/plugin-transform-runtime": {
      "optional": true
    }
  }
}
```

### duplicating babel helpers

*pending solution*

A very good feature. We have `@babel/transform-runtime` configured in our preset aswell. Our configuration differs between projects. All do use the `helpers` option and some enable `corejs` and `regenerator`. `useESModules` is always determined by `api.caller` (same way as [@babel/preset-env](https://github.com/babel/babel/blob/f8eb290da121c65566f1ce1db90de0b1a9b8e186/packages/babel-preset-env/src/index.js#L236)).

Maybe we only ensure that `@babel/transform-runtime` is included and `helpers` is set to `true`.

wessberg commented on package.json:
```diff
@@ -90,9 +82,43 @@
+		"@babel/plugin-proposal-object-rest-spread": "^7.6.2",
+		"@babel/plugin-proposal-optional-catch-binding": "^7.2.0",
+		"@babel/plugin-proposal-unicode-property-regex": "^7.6.2",
+		"@babel/plugin-syntax-dynamic-import": "^7.2.0",
```
it turns out that @babel/preset-env already depends on all of the proposal-* plugins that rollup-plugin-ts depends on, so we can leave all of these out.

wessberg commented on package.json:
```diff
@@ -90,9 +82,43 @@
+		},
+		"@babel/plugin-syntax-dynamic-import": {
+			"optional": true
+		},
```
it turns out that @babel/preset-env already depends on all of the proposal-* plugins that rollup-plugin-ts depends on, so we can leave these out.

wessberg commented:
> ### separate minification-related plugins/presets
> 
> As we do not know which plugins may get loaded for a filename `hasMinifyOptions` is always set to `true`. That currently results in some plugins applied twice (`renderChunk` and `transform`). Maybe we could change which plugins are selected during `renderChunk` (`minifyConfig`). I would suggest to only include those that have been excluded during `transform` or are syntax plugins (`babel-plugin-syntax-*` or just `[-/]syntax-*`). What do you think?

Yes, agreed. Basically, it's not so much minification vs not, but rather "what should run per-file" vs "what should run per chunk". Mangling, for example, is definitely better per chunk than per file. So, for starters, I think it makes sense to stop calling it minifyOptions and start calling it `chunkConfig` or something like that.

I agree with you on the syntax-plugins-plus-whatever-was-excluded-during-transform point.

> ### `modules` option of `preset-env` and the other yearly presets
> I included the `caller` api which is used by `preset-env` to detect if it can emit ES Modules. The user should use `modules: 'auto'` (the default value of `preset-env`). `modules: 'auto'` allows to use the same babel config for different transpile targets like rollup (esm) and jest (cjs) and it _just works_.

Sounds great!

> ### de-duplicating babel helpers
> Maybe we only ensure that `@babel/transform-runtime` is included and `helpers` is set to `true`. What do you think?

Yes, due to the reasons [outlined here](https://github.com/wessberg/rollup-plugin-ts/blob/master/documentation/explainer/why_is_babel-plugin-transform-runtime_and_tslib_included_by_default.md) and like known from other plugins like [this](https://github.com/rollup/rollup-plugin-babel#helpers), I very strongly believe it is a responsibility of this plugin to avoid helper functions being duplicated. That is the very reason for depending on `@babel/plugin-transform-runtime`. There have been few issues such as reported [here](https://github.com/wessberg/rollup-plugin-ts/issues/37) when users have depended on conflicting versions of `@babel/plugin-transform-runtime`, and I've even considered moving the helpers into the `rollup-plugin-ts` codebase itself to avoid such incompatibilities, but that comes with the price of needing to maintain them.

We may end up doing something like that, but for now, my suggestion is to keep forcing the plugin, like we do with `tslib`.

> ### Move babel plugins and presets into optional dependencies
> That would mean they get installed anyway. I think they should move into `peerDependencies` which would ensure only semver compatible versions are installed.

Originally my concern was that it would be one hell of a `npm install` command:

```
npm install @wessberg/rollup-plugin-ts @babel/preset-env @babel/plugin-proposal-async-generator-functions @babel/plugin-proposal-json-strings @babel/plugin-proposal-object-rest-spread @babel/plugin-proposal-optional-catch-binding @babel/plugin-proposal-unicode-property-regex @babel/plugin-syntax-dynamic-import @babel/plugin-transform-runtime
```

But, it turns out that `@babel/preset-env` already depends on all of the proposal-* plugins that `rollup-plugin-ts` depends on, so I've just [pushed a commit](https://github.com/wessberg/rollup-plugin-ts/commit/2666fbe18fe064e8c9a9b115475d81de940a5a27) to the master branch that removes all redundant babel dependencies and also suggested that we remove them from your PR.

sastan commented:
Sorry for the delay. I pushed a commit to remove the additional plugins.

This is my plan:

1. `minifyOptions` / `chunkConfig`: Going to implement it as suggested but keep the naming or make the default of `chunkConfig` the value of `minifyOptions` for compatibility.
2. `@babel/plugin-transform-runtime`:
  - ensure only one version is used
  - ensure helpers are referenced as imports to allow rollup to chunk them
  - ensure the referenced issues are solved by the solution

I hope i can get something done by thursday.

sastan commented:
### separate minification-related plugins/presets

*included in updated PR*

- renamed to `chunk*`
- on file transformation chunk plugins/presets are excluded
- on chunk transformation only syntax and chunk plugins/presets are included

### Move ´@babel/preset-env` into `peerDependencies`

*included in updated PR*

- marked as options via `peerDependenciesMeta`

### de-duplicating babel helpers

*included in updated PR*

- see [enforceBabelTransformRuntime](https://github.com/wessberg/rollup-plugin-ts/pull/41/files#diff-d4af9e3a6abab697e450a1d4106e5762R122)

The current implementation ensures that `@babel/plugin-transform-runtime` is included. If the user has included the plugin himself, the defaults (`FORCED_BABEL_PLUGIN_TRANSFORM_RUNTIME_OPTIONS`) are merged with his options which always have a higher precedence. For example setting `helpers` to `false` in his options would disable helper imports. I choose to do it that way because of:

1. if `@babel/plugin-transform-runtime` is not included in the user config, the default are applied 
2. `noBabelConfigCustomization` is an advanced option and user should now what they are doing 
3. overriding one option and not another maybe confusing

### Todo

- [ ] update documentation with hint that `@babel/preset-env` must be installed manually
- [ ] maybe move `@babel/plugin-transform-runtime` into `peerDependencies`
- [ ] document and explicit hint about the `@babel/plugin-transform-runtime` option merging

sastan commented:
I rebased to PR and removed `@babel/preset-env` from `peerDependencies` to prevent  breaking changes for users.

@wessberg Is there anything i can to do to get this PR merged?

wessberg commented:
Great work. I'm testing out your PR, but I'm running into a crash:

First, in the function `configItemIsChunkRelated` that now takes a string:

https://github.com/wessberg/rollup-plugin-ts/blob/2464c56495e6592b23eafa01adaeed3a5c2e13d3/src/util/get-babel-config/get-babel-config.ts#L87

Sometimes, the function is called with _not_ an ID, but a `ConfigItem`, as can be seen in the following screenshot. This may lead to crashes.

<img width="908" alt="Screen Shot 2019-11-29 at 8 05 53 PM" src="https://user-images.githubusercontent.com/20454213/69886722-a8f92d80-12e3-11ea-837b-70ea8bcd681a.png">

Currently on master, it takes a `IBabelConfigItem`:

https://github.com/wessberg/rollup-plugin-ts/blob/be1c80bc2f6944acfa863dc4aa0312ad16c991cd/src/util/get-babel-config/get-babel-config.ts#L67

It would seam that there's some normalization that needs to happen there.

sastan commented:
Sorry for the delay. I just tested a new config.

Thanks for the detailed report.

Now it should work.

wessberg commented:
Hey there. Sorry for the very late response, but the code base has diverged significantly since this PR was originally created. However, the good news is that your excellent ideas and the general approach of your work has been integrated with `rollup-plugin-ts` in the new v2.0.0 release.

- All Babel related dependencies are now optional peer dependencies
- And, `@babel/preset-env` won't be applied as a preset if the `browserslist: false` option is passed to the plugin.

Because of this, I'm closing this PR. Thanks for your great work.
